components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthnService.java

```diff
@@ -258,4 +315,149 @@ private boolean canAuthenticate(OAuthClientAuthenticator oAuthClientAuthenticato
+            //   Obtain client ID from the authorization header when basic authentication is used
+            } else if (request.getHeader(OAuthConstants.HTTP_REQ_HEADER_AUTHZ) != null) {
+                return getClientIdFromBasicAuth(request);
+            } else {
+                throw new OAuthClientAuthnException("Unable to find client id in the request.",
+                        OAuth2ErrorCodes.INVALID_CLIENT);
+            }
+        } catch (ParseException e) {
+            throw new OAuthClientAuthnException("Error occurred while parsing the signed assertion.",
+                    OAuth2ErrorCodes.INVALID_REQUEST, e);
+        }
+    }
+
+    /**
+     * Obtain the client ID of the application from the JWT in the request.
+     *
+     * @param signedObject       Client assertion sent in the request.
+     * @return Client ID of the application.
+     * @throws ParseException An exception is thrown when the signed JWT cannot be processed.
+     */
+    private String getClientIdFromJWT(String signedObject) throws ParseException {
+
+        SignedJWT signedJWT = SignedJWT.parse(signedObject);
+        return signedJWT.getJWTClaimsSet().getIssuer();
+    }
+
+    /**
+     * Obtain the client ID of the application from the authorization header of the request.
+     *
+     * @param request       Http servlet request.
+     * @return Client ID of the application.
+     * @throws OAuthClientAuthnException OAuth Client Authentication Exception.
+     */
+    private String getClientIdFromBasicAuth(HttpServletRequest request) throws OAuthClientAuthnException {
+
+        String basicAuthErrorMessage = "Unable to find client id in the request. Invalid Authorization header found.";
+        String authorizationHeader = request.getHeader(OAuthConstants.HTTP_REQ_HEADER_AUTHZ);
+        if (authorizationHeader.split(" ").length == 2) {
+            if (authorizationHeader.split(" ")[0].equals(OAuthConstants.HTTP_REQ_HEADER_AUTH_METHOD_BASIC)) {
+                String authToken = authorizationHeader.split(" ")[1];
+                byte[] decodedBytes = Base64.getUrlDecoder().decode(authToken.getBytes(StandardCharsets.UTF_8));
+                String decodedAuthToken = new String(decodedBytes, StandardCharsets.UTF_8);
+                if (decodedAuthToken.split(":").length == 2) {
+                    return decodedAuthToken.split(":")[0];
+                }
+            }
+        }
+        throw new OAuthClientAuthnException(basicAuthErrorMessage, OAuth2ErrorCodes.INVALID_CLIENT);
+    }
+
+    /**
+     * Validate whether a TLS certificate is passed through the request.
+     *
+     * @param request     Http servlet request.
+     * @return Whether a TLS certificate is passed through the request.
+     * @throws OAuthClientAuthnException OAuth Client Authentication Exception.
+     */
+    private boolean isMTLSEnforced(HttpServletRequest request) throws OAuthClientAuthnException {
```

Comment:
Need to check whether it should just be TLS and check the requirement of the validation

Reply:
Fixed in 2de86edd1bbb2cffda8c6ea07086a4746f3b1843